collect/throughput_calculator.go

```diff
@@ -0,0 +1,188 @@
+package collect
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/honeycombio/refinery/config"
+	"github.com/honeycombio/refinery/internal/peer"
+	"github.com/honeycombio/refinery/pubsub"
+	"github.com/jonboulle/clockwork"
+)
+
+const emaThroughputTopic = "ema_throughput"
+
+// EMAThroughputCalculator encapsulates the logic to calculate a throughput value using an Exponential Moving Average (EMA).
+type EMAThroughputCalculator struct {
+	Config config.Config   `inject:""`
+	Clock  clockwork.Clock `inject:""`
+	Pubsub pubsub.PubSub   `inject:""`
+	Peer   peer.Peers      `inject:""`
+
+	throughputLimit uint
+	weight          float64       // Smoothing factor for EMA
+	intervalLength  time.Duration // Length of the interval
+	hostID          string
+
+	mut         sync.RWMutex
+	throughputs map[string]throughputReport
+	clusterEMA  uint
+	lastEMA     uint // Previous EMA value
+	eventCount  int  // Internal count of events in the current interval
+	done        chan struct{}
+}
+
+// NewEMAThroughputCalculator creates a new instance of EMAThroughputCalculator.
+func (c *EMAThroughputCalculator) Start() error {
+	cfg := c.Config.GetAllSamplerRules().ThroughPutLimit
+	c.throughputLimit = uint(cfg.Limit)
+	c.done = make(chan struct{})
+
+	// if throughput limit is not set, disable the calculator
+	if c.throughputLimit == 0 {
+		return nil
+	}
+	c.intervalLength = time.Duration(cfg.AdjustmentInterval)
+	c.weight = cfg.Weight
+	c.lastEMA = 0
+
+	peerID, err := c.Peer.GetInstanceID()
+	if err != nil {
+		return err
+	}
+	c.hostID = peerID
+	c.throughputs = make(map[string]throughputReport)
+	// Subscribe to the throughput topic so we can react to throughput
+	// changes in the cluster.
+	c.Pubsub.Subscribe(context.Background(), stressReliefTopic, c.onThroughputUpdate)
+
+	go func() {
+		ticker := c.Clock.NewTicker(c.intervalLength)
```

Comment:
We should only publish if the throughput is different from the previous calculation